pulser/simulation/simulation.py

```diff
@@ -339,31 +347,31 @@ def write_samples(slot: _TimeSlot,
 
             # Case of clean global simulations
             if addr == 'Global' and not self.config.noise:
-                samples_dict = self.samples[addr][basis]
+                samples_dict = self.samples['Global'][basis]
                 if not samples_dict:
                     samples_dict = prepare_dict()
                 for slot in self._seq._schedule[channel]:
                     if isinstance(slot.type, Pulse):
                         write_samples(slot, samples_dict, True)
-                self.samples[addr][basis] = samples_dict
+                self.samples['Global'][basis] = samples_dict
 
             # Any noise : global becomes local for each qubit in the reg
             # Since coefficients are modified locally by all noises
             else:
-                is_global_pulse = addr == 'Global'
-                addr = 'Local'
+                is_global = addr == 'Global'
                 samples_dict = self.samples['Local'][basis]
                 for slot in self._seq._schedule[channel]:
                     if isinstance(slot.type, Pulse):
                         # global to local
-                        for qubit in self._qid_index:
-                            # We don't write samples for badly prep qubits
-                            if self._bad_atoms[qubit]:
-                                continue
+                        targets = (self._qid_index if is_global else
+                                   slot.targets)
```

Comment:
Actually, the time slot of a global slot also has all targets in `slot.targets`, so you don't even need this if.